### PR TITLE
[quant][refactor] Refactor find_matches for easier future extension

### PR DESCRIPTION
**Pull request by jerryzh168**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #75241
* #75202
* #75135
* #74882
* __->__ #74878

Summary:
Previously we record the matched node as a list of nodes: `List[Node]`, this does not generalize
to a graph, which is needed for future use cases, in this PR we changed the recorded node as
NodePattern instead, currently defined as
```
NodePattern = Union[Tuple[Node, Node], Tuple[Node, Tuple[Node, Node]], Any]
```
but can be more general.

This will allow us to support more general patterns with backend_config_dict api, and is also needed
for BinaryOpQuantizeHandler refactor

Test Plan:
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D35203616](https://our.internmc.facebook.com/intern/diff/D35203616)

### CODE REVIEW

**Comment by jerryzh168** [comment]
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35203616).

File: torch/ao/quantization/fx/quantization_patterns.py [hunk 1]
```diff
@@ -34,18 +43,24 @@
 class QuantizeHandler(ABC):
     """ Base handler class for the quantizer patterns
     """
-    def __init__(self, node: Node, modules: Dict[str, torch.nn.Module]):
+    def __init__(
+            self,
+            node_pattern: NodePattern,
+            modules: Dict[str, torch.nn.Module],
+            root_node_getter: Callable = None):
```
Comment (vkuzo) [inline torch/ao/quantization/fx/quantization_patterns.py:50]: what's the context to needing to override this? can this be avoided?

Comment (jerryzh168) [reply torch/ao/quantization/fx/quantization_patterns.py:50]: this is not need for current fbgemm patterns, but needed to support more complex patterns, e.g. conv - add - relu

**Comment by jerryzh168** [comment]
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35203616).

File: torch/ao/quantization/fx/quantization_patterns.py (hunk 1 above)
Comment (jerryzh168) [reply torch/ao/quantization/fx/quantization_patterns.py:50]: e.g. 
(nn.ReLU, (operator.add, (nn.BatchNorm2d, nn.Conv2d), MatchAllNode))

Comment (vkuzo) [reply torch/ao/quantization/fx/quantization_patterns.py:50]: would it make sense to add this in the future when it's needed?  I wasn't sure how this is used in this PR

**Review by vkuzo — APPROVED** [review]

File: torch/ao/quantization/fx/quantization_patterns.py (hunk 1 above)
Comment (jerryzh168) [reply torch/ao/quantization/fx/quantization_patterns.py:50]: the code path is exercised in the next PR, but I'll add a test later to test the custom root_node_getter functions, this is currently not a priority and we would like to finish the frontend refactor asap

**Comment by jerryzh168** [comment]
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35203616).

**Comment by jerryzh168** [comment]
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35203616).

**Comment by jerryzh168** [comment]
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35203616).

**Comment by jerryzh168** [comment]
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35203616).
